### PR TITLE
Refactor requests session authentication implementation

### PR DESCRIPTION
**Pull request by Obdam**
### Proposed changes
* Added an OpenCTIApiAuthentication class which extends requests.auth.AuthBase.
* This enables users of the API client to extend their requests with custom authentication.
* The old api_client repeated a couple of authentication parameters (headers, cert, verify, etc.) in each of it HTTP requests. With a custom AuthBase class, the api_client code is more DRY.
* The OpenCTIApiAuthentication class and opencti_api_client both makes sure the custom headers of applicant_id, synchronized_upsert, and retry_number can still be set when processing a queue in the Worker.

### Related issues
* None.

### Checklist
- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments
* This is an update/rework/refactor of my changes in https://github.com/OpenCTI-Platform/client-python/pull/478.

### CODE REVIEW

File: pycti/api/opencti_api_authentication.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        self.bearer_token = bearer_token
+        self.proxies = proxies
+        self.verify = verify
+        self.cert = cert
+        self.applicant_id = applicant_id
+        self.synchronized_upsert = synchronized_upsert
+        self.retry_number = retry_number
+
+    def __call__(self, r):
+        r.headers = {
```
Comment (rwblokzijl) [inline pycti/api/opencti_api_authentication.py:29]: Would make this a class variable

File: pycti/api/opencti_api_authentication.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            r.headers.update({"synchronized-upsert": self.synchronized_upsert})
+        if self.retry_number is not None:
+            r.headers.update({"opencti-retry-number": str(self.retry_number)})
+        r.proxies = self.proxies
+        r.verify = self.verify
+        r.cert = self.cert
+
+        return r
+
+    def set_applicant_id(self, applicant_id):
```
Comment (rwblokzijl) [inline pycti/api/opencti_api_authentication.py:46]: Would generalise these to `set_header`

File: pycti/api/opencti_api_authentication.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        pycti_version,
+        proxies,
+        verify,
+        cert,
+        applicant_id=None,
+        synchronized_upsert=False,
+        retry_number=None,
+    ):
+        super().__init__()
+        self.api_url = url + "/graphql"
```
Comment (rwblokzijl) [inline pycti/api/opencti_api_authentication.py:18]: Would not store this in Auth

File: pycti/api/opencti_api_client.py [hunk 4]
```diff
@@ -336,23 +338,17 @@ def query(self, query, variables={}):
                     file_index += 1
             # Send the multipart request
             r = self.session.post(
-                self.api_url,
+                url=self.auth.api_url,
```
Comment (rwblokzijl) [inline pycti/api/opencti_api_client.py:341]: Would revert this to self.api_url

**Comment by Obdam** [comment]
@SamuelHassine @richard-julien Hi Samuel and Julien, could you review my code and changes? I am curious what you think about it!

**Comment by Obdam** [comment]
@SamuelHassine @richard-julien Hi Samuel and Julien, did you had any chance to review my change? Thanks in advance!

**Comment by richard-julien** [comment]
Hi @Obdam , thanks for the PR.
Some remarks.
Can you start with rebasing the client, thanks.
Then, quite happy with the auth opencti object approach to prevent any usage errors, but not agree with the fact that this object can come from an external usage like the worker.

I understand that you need to be able to setup custom headers. So if its ok for you can you modify

**The client**
Remove the auth parameter to always instantiate the auth object in the client.
Add a custom headers parameters instead of the auth parameter

**The worker**
Remove the building of the auth object and just passthrough the custom headers to the client

Thanks

**Comment by Obdam** [comment]
Hi @richard-julien, thanks for the reply and considering this PR!
 
Let me shine more light on the purpose behind my PR. The way how I deployed the OpenCTI web application on Google Cloud (GCP) is, among other things, protected with an Identity Aware Proxy (IAP). In order for the Worker to connect to OpenCTI Web/API it needs to go through the IAP. In GCP, you [can authenticate a call through the IAP](https://cloud.google.com/run/docs/authenticating/service-to-service#acquire-token) by adding an extra authentication header.

For this I generate a proxy authentication token that needs to be set in the header. However, this token has a lifetime of 1 hour maximum. This means I need a way to set dynamic headers on the requests. 
 
The callable auth object in the `requests` library handles use cases like this. It calls the auth object on every API request. The auth object can in turn contain the logic to refresh my token.
 
It made the most sense to me to allow the generic injection of this object from the user of the client library, since this allows anyone with a similar issue to update the request object before sending it. I can then fork the worker and add the GCP functionality, or create a PR to add support for it depending on some environenment variables.
 
I hope you share my interest in supporting some form of dynamic header setting.
 
Thank you for your consideration.

**Comment by richard-julien** [comment]
> Hi @richard-julien, thanks for the reply and considering this PR!
> 
> Let me shine more light on the purpose behind my PR. The way how I deployed the OpenCTI web application on Google Cloud (GCP) is, among other things, protected with an Identity Aware Proxy (IAP). In order for the Worker to connect to OpenCTI Web/API it needs to go through the IAP. In GCP, you [can authenticate a call through the IAP](https://cloud.google.com/run/docs/authenticating/service-to-service#acquire-token) by adding an extra authentication header.
> 
> For this I generate a proxy authentication token that needs to be set in the header. However, this token has a lifetime of 1 hour maximum. This means I need a way to set dynamic headers on the requests.
> 
> The callable auth object in the `requests` library handles use cases like this. It calls the auth object on every API request. The auth object can in turn contain the logic to refresh my token.
> 
> It made the most sense to me to allow the generic injection of this object from the user of the client library, since this allows anyone with a similar issue to update the request object before sending it. I can then fork the worker and add the GCP functionality, or create a PR to add support for it depending on some environenment variables.
> 
> I hope you share my interest in supporting some form of dynamic header setting.
> 
> Thank you for your consideration.

Hi @Obdam , sorry for the delay.

Thanks for the explanation, perfectly fine for me to support more kind of authentication.
For the PR, as for now the addition is really to be able to add custom headers I would prefer to only support this use case instead of a complete override of the auth object. Depending of future use case it could be useful but for now limiting the possibilities looks like a better approach short term.

Do you think you could consider reworking a bit this approach to only passthrough the custom headers?

**Pull request closed without merging**